Proxy authentication support (eq. oauth2-proxy)

nielscil opened the pull request:
**Introduction**
When searching for SSO support in the issues, I came across https://github.com/anonaddy/anonaddy/issues/248#issuecomment-1407264756 where is being discussed to support proxy authentication like oauth2-proxy. @Queuecumber would make a PR for this, but I couldn't find it, so I made it myself.

**Changes in this PR**
- Moved redirect logic from `LoginController` to helper class.
- Moved user creation logic from `RegisterController` helper class.
- Added config/env entries for:
    - enabling proxy (`use_proxy_authentication` and `ANONADDY_USE_PROXY_AUTHENTICATION`).
    - User header naming (`proxy_authentication_username_header` and `ANONADDY_PROXY_AUTHENTICATION_NAME_HEADER`).
    - Email header naming (`proxy_authentication_email_header` and `ANONADDY_PROXY_AUTHENTICATION_EMAIL_HEADER`).
-  Added Middleware `ProxyAuthentication`.
- Added  `ProxyAuthenticationTest` for testing the middleware.

**Proxy Authentication flow**
- Check if proxy is enabled, if not; go to next middleware.
- Check if currently logged in (using proxy or normal login flow).
- If there is a change (other proxy user, normal flow to proxy flow or no proxy headers anymore), the current user is logged out.
- If there are headers provided.
    - create user when not in db.
    - login.
 - if logged in, redirect to right location.
 - if logged out, redirect to login page.
 - else go to next middleware.

willbrowningme commented:
Thanks for the PR! Is anyone able to test and verify this?

Queuecumber commented:
I'd love to try it but I'm not sure when I will get time, I will try to prioritize it

Queuecumber commented:
> If there are headers provided.
>   *    create user when not in db.
>   *   login.

Is there a way to disable this? I can see it being abused for private servers

nielscil commented:
Isn't it the whole point that the headers register/login? The Proxy should authenticate and proxy the right user and email to Addy. the headers behaviour can be toggled  with `use_proxy_authentication`. Default is `false`, which uses normal login flow. `true` uses the header checking to login or register and login.

Queuecumber commented:
Yeah actually that makes sense, since the proxy would be managing who is allowed to log by the time it hits Addy we can assume that either 1) the user exists or 2) we should provision it because the external auth approved the user

I think it's fine as is then

nielscil commented:
So I did some testing on my production setup and encountered some functional things:
- All items on the security settings page won't work because it needs the verification of current password. Not an issue per se, but bit confusing. Should we disable this page when using proxy authentication?
- Update email and delete account setting won't work either because of this verification.
- Logout button isn't gonna work because we have no knowledge of the authentication proxy. Should we disable it?
- API key creation cannot be done without verification of current password. This is an real issue because this way no api clients can be used. For proxy authentication this should be without verification of current password.

Queuecumber commented:
Yeah I would vote to disable any additional authentication screens when using proxy auth

I'm planning to deploy this today do you have a docker image handy for it?

nielscil commented:
> I'm planning to deploy this today do you have a docker image handy for it?

I created a docker image for testing: ghcr.io/nielscil/anonaddy-docker:edge

nielscil commented:
So, I've made the following changes when using proxy authentication:

- Logout button removed.
- skipping 2fa.
- disable security and account deletion settings tabs.
- create api tokens without password.

I'm not sure what to do with the 'Update Email' option, because the email is provided from the proxy. I see two possible solutions:

1. Remove option to update email entirely and let the middleware check and update the email on login.
2. Remove password verification and allow users to deviate from the provided email.

Queuecumber commented:
Deployed and working well so far, will play with it a bit and let you know but also feel free to suggest some specific things to test

So far: 
* Tested logging in with new user
* Testing generating API key
* Tested using API key for generating alias

Queuecumber commented:
Actually one thing:

* Tested changing user name header, creates a new user in the database

I think this is probably the correct behavior, just noting it down

If the `ANONADDY_PROXY_AUTHENTICATION_NAME_HEADER` changes it creates a new user in the database rather than updating the user with the matching `ANONADDY_PROXY_AUTHENTICATION_EMAIL_HEADER`

nielscil commented:
I've been thinking about that but is there a provided header that stays the same when changing a username? something like a userid? I'm using Authentik to proxy and it has a `X-authentik-uid` header. If so, we could store the value of this header in a Username and use it to check instead of the username itself. 

I'm also struggling with the option to add usernames from UI. Now you could possible add a username, which the upstream authentication provider uses for another user. When that other user logs into Addy, it then goes to the first user instead of creating a new user.
The problem here is that I can hide adding usernames from web UI, but it's also possible from the API to do (e.q. in the mobile app).

My suggestion would be:

1. Add an external flag to the `username` entity and block all username add/delete/canlogin actions for users with a username with that flag on true.
2. Add an external id to the `username` entity, use this id to register + login.

The second option looks better to me and keeps username functionality in Addy as makes it possible to rename username when proxy changes it but in that case we need a unique identifier.

Queuecumber commented:
Yeah this is tricky because it expands the scope of the MR, but I think you would have to implement option (2)

My setup needs the ability to add usernames because I use google authentication which doesn't send anything reasonable in the headers that set up the default username (it's either the email address in `X-Auth-Request-Email` or some kind of internal numerical ID in `X-Auth-Request-User`). 

So we need to manually add a user friendly username for catch-all

Queuecumber commented:
Could this be as simple as force disabling `canlogin` for any usernames added via the UI? 

The only caveat is that you could block someone by adding their username in your UI but that isnt a problem which is specific to the feature of this MR right?

nielscil commented:
> Could this be as simple as force disabling `canlogin` for any usernames added via the UI? 

When proxy is enabled, block all calls to canlogin. I guess that can be done without having an external id. 

> The only caveat is that you could block someone by adding their username in your UI but that isnt a problem which is specific to the feature of this MR right? 

Normally the application gives an error when username is taken and the user can choose something else. Now the proxy won't login.

Queuecumber commented:
I wanted to update that I've been running this in "production" all week and its been solid

nielscil commented:
I made last week some changes, mostly finetuning of UI, blocking things we don't want if we're using proxy authentication and tests. I've updated the PR to reflect this changes. 

The biggest change is that this PR (and the testing docker image) now using a extra Header for providing an external id that is saved in the database by an username. therfore it has a migration now. **If testing, please backup your data!**
I've added a command for mapping existing usernames to external ids:
`anonaddy:set-username-externalid <username> <id>`

If, after testing, you would like to go back to the official (docker) image, don't forget to remove the migration using:
`php artisan migrate:rollback`

Queuecumber commented:
Hey how are we feeling out this? I've been using it without issue

willbrowningme commented:
Has anyone else tested this in production? It's not something I'll be using, I'm curious to know how many people will actually make use of this.

Queuecumber commented:
> I made last week some changes, mostly finetuning of UI, blocking things we don't want if we're using proxy authentication and tests. I've updated the PR to reflect this changes.

I had actually missed this comment so I just now pulled the latest image, I'm not able to log in with it. After I do my SSO I get dumped at the standard anonaddy login page, did something change with the environments that I need in order to activate proxy auth?

> The biggest change is that this PR (and the testing docker image) now using a extra Header for providing an external id that is saved in the database by an username

What was the motivation for going with the external id over `canlogin`? Would it prevent the scenario where a malicious user blocks another user by registering their username?

nielscil commented:
You need 3 headers (I updated original PR message to reflect this). The id should be unique and not changing for an specific user.

You also need to mark existing users with this id by using the command in my previous message, otherwise an new user is created.

 It's indeed to prevent other users from accessing existing accounts when they use (or change to) the same details and for the api to know if an user is external to block specific actions.

Queuecumber commented:
> You need 3 headers 
Yes I added the new header and even redid my database but I was getting 403 forbidden with my previous setup

It turned out that it was caused by 

```
ANONADDY_PROXY_AUTHENTICATION_NAME_HEADER: X-Auth-Request-Email
```

which I think hits this? 

https://github.com/anonaddy/anonaddy/pull/695/files#diff-4149d6e6d9635f2f4949cdc4b6ffc473ffd8cc89259c23d5768a59c052d60f4fR125 

it's very prescriptive on how usernames should look

With Google auth you have two options: 

1. The email address (email claim, mapped to X-Auth-Request-Email)
2. A very long meaningless number (sub claim, mapped to X-Auth-Request-User)

These are the only two claims you are guaranteed, so with the new setup our usernames in the top left are these long meaningless numbers, previously I had them set to the email addresses so they would display nicely and then we used custom usernames to make workable catchall aliases

> It's indeed to prevent other users from accessing existing accounts when they use (or change to) the same details and for the api to know if an user is external to block specific actions.

Can you elaborate on this a little? Like what actions would be available to a malicious user which this prevents and what actions are blocked? 

It does add quite a bit to the MR which now touches database schema and things so I think it should probably be tested fairly extensively without proxy auth to make sure everything still works

nielscil commented:
Ah yes, it follows the default username rules where it previously didn't. I don't think that should be reversed because addy has that as requirement. I could add some code to take only the localpart of an email.

The database change adds a extra field only used in the oauth middleware. It checks which user to login and if not exist creates a new user and makes a new available username for it. The normal flow doesn't use or fill it.

In the previous situation it was possible that an oauth provider changed the username to one that was already in use by anonaddy and by that way got control over an account that it should not have. New situation only uses the id, which should not change when an username changes

Queuecumber commented:
> Ah yes, it follows the default username rules where it previously didn't. 

OK if its following the existing username rules then I think it probably makes sense to keep it

> I could add some code to take only the localpart of an email

This is a good idea but again increases scope

Other comments regarding external id make sense and so far things seem to be working

Queuecumber commented:
Any news on this? I'm still using it and it still works well

The previous issue of my default username being a long string of numbers because of google seems to be a pretty minor issue, I think if it really needs solving it can be revisiting in a future PR

xairam commented:
Is there any ETA of implementing the merge in master?

I just wondered if this would work with Authelia, since I am using lldap/Authelia for my families other services(paperless, mealie, etc.)

nielscil commented:
The PR is done on my end. I have it running for months without problems. @willbrowningme any idea when it would be possible to merge this?

If Authelia supports OpenID Connect (OIDC) you could use Addy behind an oauth2-proxy, which authenticates against your Authelia instance. I can't give you a tutorial on how to do that because I'm using Authentik's build-in proxy for this.

xairam commented:
This should work for authentication side right?

gonna give it try with your container tonight.

**nginx conf:**
`location / {
      
        auth_request /authelia-auth;

        
        auth_request_set $user $upstream_http_remote_user;
        auth_request_set $email $upstream_http_remote_email;
        auth_request_set $uid $upstream_http_remote_user_id;

        
        proxy_set_header Remote-User-Id $uid;
        proxy_set_header Remote-Username $user;
        proxy_set_header Remote-Email $email;

        proxy_pass http://anonaddy:8000;
    }`

willbrowningme commented:
Is anyone able to give me instructions on how to test this out myself?

nielscil commented:
@xairam I'm sorry, but can't help you with that. Like I said, I'm not using Authelia (so I don't know hot that works) and I'm using traefik as reverse proxy (not for addy, I'm using Authentik build-in authentication proxy for that). 

Did you manage to get it to work?

@willbrowningme Sorry for the late response, been a bit busy. 

For just testing if Addy works, I'm using [ModHeader extension](https://chromewebstore.google.com/detail/modheader-modify-http-hea/idgpnmonknjnojddfkpgkljpfnnfcklj?pli=1). Don't do this on any public facing instance because anyone can add the headers.

preperations:
- Set `ANONADDY_USE_PROXY_AUTHENTICATION` to `true` in env file.
- Set `ANONADDY_PROXY_AUTHENTICATION_USER_ID_HEADER` to `X-User` in env file.
- Set `ANONADDY_PROXY_AUTHENTICATION_NAME_HEADER` to `X-Name` in env file.
- Set `ANONADDY_PROXY_AUTHENTICATION_EMAIL_HEADER` to `X-Email` in env file.

In ModHeader, add a profile, add the headers to the request and go to the addy instance. It should create a new user for the provided values in the headers.

----
I cannot give you instructions for the whole stack (that is, with oauth2-proxy), because I'm not using that. @Queuecumber can you provide some config for that?

Queuecumber commented:
Let me see if I can throw something together with docker compose so we can get this moving again

nielscil commented:
I merged upstream into my branch. 

@willbrowningme do you know if database migrations are run in order? My migration is now not the latest anymore, so I don't know if the database updater skips this migration because there is already a newer migration executed.

willbrowningme commented:
@nielscil your migration will still be run as it won't yet be in the migrations table.

buxm commented:
Any idea if this can be made to work with Cloudflare Tunnels (`cloudflared`) and Cloudflare Access?

I can set up a Cloudflare tunnel on a `cloudflared` daemon local to my `anonaddy` server. The tunnel is reachable on a specific subdomain (e.g. `addy.mydomain.com`) and is protected by a Cloudflare Access policy that requires logging in through an integrated IdP (e.g. Google).
Only when a user successfully logs in and has a specific email address, `cloudflared` would reverse proxy the request to `anonaddy`.
However the only HTTP headers `cloudflared` sends through to `anonaddy` that can be used to identify the user are:
`Cf-Access-Authenticated-User-Email`: the email address of the user logged in through the IdP
`Cf-Access-Jwt-Assertion`: a Jwt containing more encrypted data about the user.

I set:
```
ANONADDY_USE_PROXY_AUTHENTICATION=true
ANONADDY_PROXY_AUTHENTICATION_USER_ID_HEADER=Cf-Access-Authenticated-User-Email
```
Then I ran:
```
anonaddy anonaddy:set-username-externalid <my-existing-anonaddy-username> <content-of-Cf-Access-Authenticated-User-Email>
```
I specifically checked the content of `Cf-Access-Authenticated-User-Email` using `httpbin` and passed it to the `anonaddy:set-username-externalid` command.

I ran the following query on the DB:
```
select external_id,username,can_login
from usernames un join users u on un.user_id=u.id
where username='<my-existing-anonaddy-username>';

+-------------------------------------------------+--------------------------------+-----------+
| external_id                                     | username                       | can_login |
+-------------------------------------------------+--------------------------------+-----------+
| <content-of-Cf-Access-Authenticated-User-Email> | <my-existing-anonaddy-username>|         1 |
+-------------------------------------------------+--------------------------------+-----------+
```

however, after navigating to `addy.mydomain.com` and logging in to the IdP, I am presented with the anonaddy login page and I am not logged in.
